### PR TITLE
Single lexer parser

### PR DESCRIPTION
**Pull request by mchernigin**
Draft for now. A lot can be improved.

### CODE REVIEW

File: ast_calc.c [hunk 1]
```diff
@@ -1,94 +1,33 @@
...
-    if (astlex_init_extra (&ast, &scanner)) {
+    if (yylex_init_extra (&ast, &scanner)) {
         fprintf (stderr, "ERROR: cannot initialize scanner\n");
         return (EXIT_FAILURE);
     }
 
-    if (ast_scan_string (config->expr, scanner) == NULL) {
+    if (yy_scan_string (config->expr, scanner) == NULL) {
         fprintf (stderr, "ERROR: cannot scan given string\n");
         return (EXIT_FAILURE);
```
Comment (hnarkaytis) [inline ast_calc.c]: Memory leak. Need to make ast_free

File: ast_calc.c [hunk 2]
```diff
@@ -1,94 +1,33 @@
...
     }
 
-    if (astparse (scanner)) {
+    if (ast_parse (scanner)) {
         fprintf (stderr, "ERROR: cannot parse string\n");
         return (EXIT_FAILURE);
```
Comment (hnarkaytis) [inline ast_calc.c]: Another memory leak

File: basic_calc.c [hunk 3]
```diff
@@ -1,28 +1,41 @@
...
-    if (basiclex_init_extra (&config->result, &scanner)) {
+    if (yylex_init_extra (&config->result, &scanner)) {
         fprintf (stderr, "ERROR: cannot initialize scanner\n");
         return (EXIT_FAILURE);
     }
 
     for (size_t i = 0; i < config->iteration_number; ++i) {
-        if (basic_scan_string (config->expr, scanner) == NULL) {
+        if (yy_scan_string (config->expr, scanner) == NULL) {
             fprintf (stderr, "ERROR: cannot set string to parse\n");
```
Comment (hnarkaytis) [inline basic_calc.c:29]: Mem leak

File: basic_calc.c [hunk 4]
```diff
@@ -1,28 +1,41 @@
...
             return (EXIT_FAILURE);
         }
 
-        if (basicparse (scanner)) {
+        if (basic_parse (scanner)) {
             fprintf (stderr, "ERROR: cannot parse string\n");
```
Comment (hnarkaytis) [inline basic_calc.c:35]: mem leak

**Review by hnarkaytis — APPROVED** [review]
LGTM

**Comment by mchernigin** [comment]
> LGTM

but not to me: segfault on linux in ast mode

File: ast_calc.c (hunk 1 above)
Comment (mchernigin) [reply ast_calc.c]: You mean inside if statement? There is an ast_free call in the end of the function.

**Review by hnarkaytis — APPROVED** [review]

**Comment by mchernigin** [comment]
I have no idea what just happened… I did not change anything in the project since last month. But for some reason, it now works both on Linux and MacOS... I guess, I'm merging...

**Pull request merged**